### PR TITLE
Introduce the ability to declare metadata using named exports

### PR DESCRIPTION
**Pull request by JosXa**
Instead of comments at the top of the file, users can now use an export named `metadata` in their scripts. If found, this will be merged with the comment-based metadata. The advantage is that metadata can be declared in a type-safe manner.

### CODE REVIEW

File: src/types/core.d.ts [hunk 1]
```diff
@@ -450,8 +450,60 @@ export type PromptConfig = {
...
 >
 
+export type CronExpression =
+  | `${string} ${string} ${string} ${string} ${string}`
+  | `${string} ${string} ${string} ${string} ${string} ${string}`
+
 export interface Metadata {
-  [key: string]: string
+  /** Specifies the name of the script as it appears in the Script Kit interface. */
+  name: string
```
[the remark this answers is not in the document]
Comment (JosXa) [reply src/types/core.d.ts:459]: ah, good to know

File: src/cli/kenv-create.ts [hunk 2]
```diff
@@ -95,21 +95,23 @@ Additional kenvs are located:
     },
   })
 
+  const kenvTemplateRepo = process.env.KENV_TEMPLATE_REPO ?? `johnlindquist/kenv-template`
```
Comment (JosXa) [inline src/cli/kenv-create.ts:98]: oops this shouldn't be here

File: src/cli/new.ts [hunk 3]
```diff
@@ -123,15 +124,30 @@ if (!arg?.tip) {
...
+  if (
+    (scriptName || command !== name) &&
+    !contents.includes(`Name:`)
+  ) {
+    contents = `// Name: ${scriptName || name || ""}
 ${contents.startsWith("/") ? contents : "\n" + contents}
 `
+  }
+} else if (defaultMetadataMode === 'convention') {
+  // TODO: Differentiate the export based on the JS/TS default setting and use jsdoc annotation
```
Comment (JosXa) [inline src/cli/new.ts:139]: [ ] Differentiate the export based on the JS/TS default setting and use jsdoc annotation

File: src/core/utils.ts [hunk 4]
```diff
@@ -458,17 +459,122 @@ export let getMetadata = (contents: string): Metadata => {
...
+  if (!/(const|var|let) metadata/g.test(contents)) {
+    // No named export in file, return early
+    return fromComments
+  }
+
+  let ast: Program
+  try {
+    ast = parseTypeScript(contents)
+  } catch(err) {
+    // TODO: May wanna introduce some error handling here. In my script version, I automatically added an error
```
Comment (JosXa) [inline src/core/utils.ts:563]: @johnlindquist what do you think?

File: src/main/kit-windows.ts [hunk 5]
```diff
@@ -28,7 +27,7 @@ await widget(md(\`# Hello world\`), {
       placeholder: "Focus Kit Window",
       enter: "Focus",
     },
-    windows as Choice[]
+    windows as Choices<any>
```
Comment (JosXa) [inline src/main/kit-windows.ts:30]: 🤷

File: src/setup/cache-grouped-scripts.ts [hunk 6]
```diff
@@ -9,15 +8,16 @@ import {
...
 let groupedScripts = await getGroupedScripts()
 let scripts = formatChoices(groupedScripts)
 let firstScript = scripts.find(
   script => !script.skip
-) as Script
+)
 let preview = ``
 try {
-  preview = await processScriptPreview(firstScript)()
+  preview = await processScriptPreview(firstScript as unknown as Script)()
```
Comment (JosXa) [inline src/setup/cache-grouped-scripts.ts:20]: - [ ] Should have another look here, apparently I broke something but idk how

File: src/types/index.d.ts [hunk 7]
```diff
@@ -13,7 +13,10 @@ export type GlobalApi = Omit<GlobalsApi, "path"> &
   ProAPI
 
 declare global {
```
Comment (JosXa) [inline src/types/index.d.ts:15]: weird that these are checked in

**Review by johnlindquist — APPROVED** [review]

File: src/types/index.d.ts (hunk 7 above)
Comment (JosXa) [reply src/types/index.d.ts:15]: Oh, okay I guess that's something we have to live with. Or automate away

**Pull request merged**